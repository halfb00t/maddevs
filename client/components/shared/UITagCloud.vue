<template>
  <div
    v-if="tagsCategories.length"
    class="posts-filter"
  >
    <ul class="posts-filter__list">
      <li
        v-for="(tag, i) in tagsCategories"
        :key="`posts-filter__item-${i}`"
        class="posts-filter__item-wrapper"
      >
        <div class="posts-filter__item">
          <input
            :id="tag.tagName"
            type="checkbox"
            name="Tag"
            data-testid="test-post-input"
            :value="tag.tagName"
            :checked="activeTags.includes(tag.tagName)"
            @change="handleFilterChange"
          >
          <label :for="tag.tagName">{{ tag.tagName }}</label>
        </div>
      </li>
    </ul>
  </div>
</template>
<script>
import { mapActions, mapGetters } from 'vuex'

export default {
  name: 'UITagCloud',

  computed: {
    ...mapGetters([
      'tags',
      'activeTags',
    ]),

    tagsCategories() {
      const { tags } = this
      if (Array.isArray(tags) && tags.length) return tags
      return []
    },
  },

  mounted() {
    this.getTagsFromPrismic()
  },

  destroyed() {
    this.clearAllData()
  },

  methods: {
    ...mapActions([
      'changeActiveTags',
      'deleteActiveTag',
      'getTagsFromPrismic',
      'clearAllData',
    ]),

    handleFilterChange(e) {
      if (this.activeTags.includes(e.currentTarget.value)) {
        return this.deleteActiveTag(e.currentTarget.value)
      }
      return this.changeActiveTags(e.currentTarget.value)
    },
  },
}
</script>
<style lang="scss" scoped>
.posts-filter {
  min-width: 150px;
  margin-bottom: 50px;

  &__list {
    display: flex;
    flex-flow: row wrap;
    justify-content: center;
    gap: 12px;
  }

  &__item {
    &-wrapper {
      &:last-child {
        margin-right: 0;
      }
    }

    input[type="checkbox"] {
      display: none;
    }

    label {
      @include font('Inter', 12px, 400);
      cursor: pointer;
      box-shadow: none;
      display: flex;
      align-items: flex-end;
      padding: 10px 16px;
      color: $text-color--black;
      background-color: $bgcolor--silver;
      border: 1px solid $border-color--grey-20-percent;
      line-height: 18px;
      border-radius: 20px;
      transition: 0.2s;
    }

    input[type="checkbox"]:checked + label {
      border-color: $border-color--red;
      color: $text-color--red;
    }
  }

  @media only screen and (min-width: 1024px) {
    &__item label:hover {
      border-color: $border-color--red;
      color: $text-color--red;
    }
  }

  @media only screen and (min-width: 991px) {
    margin-bottom: 35px;
  }

  @media screen and (max-width: 460px) {
    &__list {
      justify-content: flex-start;
    }
  }
}
</style>
