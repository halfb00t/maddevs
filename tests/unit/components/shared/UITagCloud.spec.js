import { render } from '@testing-library/vue'
import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import UITagCloud from '@/components/shared/UITagCloud'

const localVue = createLocalVue()
localVue.use(Vuex)

const getters = {
  tags: () => [{ tagName: 'Editors pick' }, { tagName: 'Python' }],
  activeTags: () => ['Python'],
}

const actions = {
  changeActiveTags: jest.fn(),
  deleteActiveTag: jest.fn(),
  getTagsFromPrismic: jest.fn(),
  clearAllData: jest.fn(),
}

let store = new Vuex.Store({
  getters,
  actions,
})

describe('UITagCloud component', () => {
  it('should correctly render component', async () => {
    const { container } = render(UITagCloud, {
      localVue,
      store,
    })

    expect(container).toMatchSnapshot()
  })

  it('should correctly work handleInputChange', async () => {
    const wrapper = await mount(UITagCloud, {
      localVue,
      store,
    })

    const tagInput = wrapper.findAll('input')
    await tagInput.at(0).trigger('change')
    expect(actions.changeActiveTags).toHaveBeenCalledTimes(1)
    await tagInput.at(1).trigger('change')
    expect(actions.deleteActiveTag).toHaveBeenCalledTimes(1)
  })

  it('should correctly return empty from tagsCategories computed', () => {
    getters.tags = () => []

    store = new Vuex.Store({
      getters,
      actions,
    })
    const wrapper = mount(UITagCloud, {
      localVue,
      store,
    })

    expect(wrapper.vm.tagsCategories).toHaveLength(0)
  })
})
