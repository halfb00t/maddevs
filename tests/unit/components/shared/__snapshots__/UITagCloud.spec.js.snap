// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UITagCloud component should correctly render component 1`] = `
<div>
  <div
    class="posts-filter"
  >
    <ul
      class="posts-filter__list"
    >
      <li
        class="posts-filter__item-wrapper"
      >
        <div
          class="posts-filter__item"
        >
          <input
            data-testid="test-post-input"
            id="Editors pick"
            name="Tag"
            type="checkbox"
            value="Editors pick"
          />
           
          <label
            for="Editors pick"
          >
            Editors pick
          </label>
        </div>
      </li>
      <li
        class="posts-filter__item-wrapper"
      >
        <div
          class="posts-filter__item"
        >
          <input
            data-testid="test-post-input"
            id="Python"
            name="Tag"
            type="checkbox"
            value="Python"
          />
           
          <label
            for="Python"
          >
            Python
          </label>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
